Source/WebCore/dom/SpeculationRulesMatcher.cpp

```diff
@@ -104,23 +104,28 @@ static bool matches(const SpeculationRules::DocumentPredicate& predicate, Docume
     );
 }
 
-// https://wicg.github.io/nav-speculation/speculation-rules.html#document-rule-predicate-matching
 std::optional<PrefetchRule> SpeculationRulesMatcher::hasMatchingRule(Document& document, HTMLAnchorElement& anchor)
 {
     const auto& speculationRules = document.speculationRules();
     const auto& url = anchor.href();
 
-    for (const auto& rule : speculationRules->prefetchRules()) {
+    std::optional<PrefetchRule> result;
+    speculationRules->forEachPrefetchRule([&result, &document, &anchor, &url](const SpeculationRules::Rule& rule) {
+        if (result)
+            return;
+
         for (const auto& href : rule.urls) {
-            if (href == url)
-                return PrefetchRule { rule.tags, rule.referrerPolicy, rule.eagerness == SpeculationRules::Eagerness::Conservative };
+            if (href == url) {
+                result = PrefetchRule { rule.tags, rule.referrerPolicy, rule.eagerness == SpeculationRules::Eagerness::Conservative };
+                return;
+            }
         }
 
         if (rule.predicate && matches(rule.predicate.value(), document, anchor))
-            return PrefetchRule { rule.tags, rule.referrerPolicy, rule.eagerness == SpeculationRules::Eagerness::Conservative || rule.eagerness == SpeculationRules::Eagerness::Moderate };
-    }
+            result = PrefetchRule { rule.tags, rule.referrerPolicy, rule.eagerness == SpeculationRules::Eagerness::Conservative || rule.eagerness == SpeculationRules::Eagerness::Moderate };
```

Comment:
I do like the early return instead of the optional, capture, and quick iteration.  Would it be reasonable to keep track of prefetch rules by Node instead of by Element, then use the Document as the Node "owning" the rules from headers?

Reply:
Done!